src/datasources/balances-api/zerion-balances-api.service.ts

```diff
@@ -0,0 +1,155 @@
+
+@Injectable()
+export class ZerionBalancesApi implements IBalancesApi {
+  private readonly apiKey: string | undefined;
+  private readonly baseUri: string;
+  private readonly chainsConfiguration: Record<number, ChainAttributes>;
+  private readonly defaultExpirationTimeInSeconds: number;
+  private readonly defaultNotFoundExpirationTimeSeconds: number;
+
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+    private readonly dataSource: CacheFirstDataSource,
+  ) {
+    this.apiKey = this.configurationService.get<string>(
+      'balances.providers.zerion.apiKey',
+    );
+    this.baseUri = this.configurationService.getOrThrow<string>(
+      'balances.providers.zerion.baseUri',
+    );
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+    this.chainsConfiguration = this.configurationService.getOrThrow<
+      Record<number, ChainAttributes>
+    >('balances.providers.zerion.chains');
+  }
+
+  async getBalances(args: {
+    chainId: string;
+    safeAddress: string;
+    fiatCode: string;
+  }): Promise<Balance[]> {
+    try {
+      const cacheDir = CacheRouter.getZerionBalancesCacheDir(args);
+      const chainName = this._getChainName(args.chainId);
+      const currency = args.fiatCode.toLowerCase();
+      const url = `${this.baseUri}/v1/wallets/${args.safeAddress}/positions?filter[chain_ids]=${chainName}&currency=${currency}&sort=value`;
+      const { data } = await this.dataSource.get<ZerionBalances>({
+        cacheDir,
+        url,
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        networkRequest: { headers: { Authorization: `Basic ${this.apiKey}` } },
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return this._mapBalances(chainName, data);
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${args.safeAddress} balances from provider: ${asError(error).message}}`,
+      );
+    }
+  }
+
+  private _mapBalances(
+    chainName: string,
+    zerionBalances: ZerionBalance[],
+  ): Balance[] {
+    return zerionBalances
+      .filter((zb) => zb.attributes.flags.displayable === true)
+      .map((zb) => {
+        const implementation = zb.attributes.fungible_info.implementations.find(
+          (implementation) => implementation.chain_id === chainName,
+        );
+        if (!implementation)
+          throw Error(
+            `Zerion error: ${chainName} implementation not found for balance ${zb.id}`,
+          );
+        const fiatBalance = getNumberString(zb.attributes.value ?? 0);
```

Comment:
Having `0` as a fallback can be dangerous as `0` is a valid value. What about skipping items with 0 value?

Reply:
I think skipping 0-value items is not an option as we would skip SAFE atm for example. But instead of returning 0 from the data-source layer, we can propagate `null` from here and let the route class map it to 0 (it is doing it in https://github.com/safe-global/safe-client-gateway/blob/d37d2b9f58b8ed28950623131da0e8f76ce86470/src/routes/balances/balances.service.ts#L78)

I've changed the fallback in [2b91828](https://github.com/safe-global/safe-client-gateway/pull/1096/commits/2b91828fa16d3a9a478461eb961910310bc11a5f)

Hopefully, we would be able to simplify the code a little if a unique provider is used in the future.